featuretools/synthesis/deep_feature_synthesis.py

```diff
@@ -517,21 +517,29 @@ def _build_transform_features(self, all_features, entity, max_depth=0,
                                                         entity,
                                                         new_max_depth,
                                                         input_types,
-                                                        groupby_prim,
-                                                        require_direct_input=require_direct_input)
+                                                        groupby_prim)
             # get IDs to use as groupby
             id_matches = self._features_by_type(all_features=all_features,
                                                 entity=entity,
                                                 max_depth=new_max_depth,
                                                 variable_type=set([Id]))
-            for matching_input in matching_inputs:
+            # If require_direct_input, require a DirectFeature in input or as a
+            # groupby, and don't create features of inputs/groupbys which are
+            # all direct features with the same relationship path
+            matches_with_groupby = set()
+            for id in id_matches:
+                for match in matching_inputs:
+                    if not require_direct_input or (
+                            not _all_direct_and_same_path(match + (id,)) and
+                            any([isinstance(m, DirectFeature) for m in (match + (id,))])):
```

Comment:
I think we instead creating another matches variable we could handle the require_direct_input check and creating a feature in the same loop and just move to the next iteration if the check fails